include/rocksdb/metadata.h

```diff
@@ -16,24 +16,105 @@
 
 namespace ROCKSDB_NAMESPACE {
 
+// Basic identifiers and metadata for a file in a DB. This only includes
+// information considered relevant for taking backups, checkpoints, or other
+// services relating to DB file storage.
+struct FileStorageInfo {
+  // The name of the file within its directory (e.g. "123456.sst")
+  std::string relative_filename;
+  // The directory containing the file, without a trailing '/'. This could be
+  // a DB path, wal_dir, etc.
+  std::string directory;
+
+  // The id of the file within a single DB. Set to 0 if the file does not have
+  // a number (e.g. CURRENT)
+  uint64_t file_number = 0;
+  // The type of the file as part of a DB.
+  FileType file_type = kTempFile;
+
+  // File size in bytes. See also `trim_to_size`.
+  uint64_t size = 0;
+  // If true, the file on disk is allowed to be larger than `size` but only
+  // the first `size` bytes should be used for the current context. If false,
+  // the file is corrupt if size on disk does not equal `size`.
+  //
+  // If a CURRENT file needs to be regenerated to have a functioning DB, its
+  // size=0 and trim_to_size=true.
+  bool trim_to_size = false;
+
+  // This feature is experimental and subject to change.
+  Temperature temperature = Temperature::kUnknown;
+
+  // The checksum of a SST file, the value is decided by the file content and
+  // the checksum algorithm used for this SST file. The checksum function is
+  // identified by the file_checksum_func_name. If the checksum function is
+  // not specified, file_checksum is "0" by default.
+  std::string file_checksum;
+
+  // The name of the checksum function used to generate the file checksum
+  // value. If file checksum is not enabled (e.g., sst_file_checksum_func is
+  // null), file_checksum_func_name is UnknownFileChecksumFuncName, which is
+  // "Unknown".
+  std::string file_checksum_func_name;
+};
+
+namespace detail {
+struct SstFileMetaDataCompat : public FileStorageInfo {
```

Comment:
Can you explain why this struct? Users who consume SstFileMetaData may feel it hard to fine the information they need. They need to track two level of inheritance just to find file name and directory, etc. It might be confusing to users. Also, I'm not sure this way of using inheritance is a good practice for OOP programming. Google C++ Style suggests "restrict use of inheritance to the "is-a" case: Bar subclasses Foo if it can reasonably be said that Bar "is a kind of" Fo" https://google.github.io/styleguide/cppguide.html#Inheritance and I think here it is not the case. I feel the original SstFileMetaData looks better.

Reply:
```
// The metadata that describes a SST file.
// FileStorageInfo is the intended base class.
struct SstFileMetaData : public detail::SstFileMetaDataCompat {
```

`FileStorageInfo` is the intended base class. SstFileMetaData is a FileStorageInfo, because it provides that information about files and more. For API compatibility, there are additional members in `detail::SstFileMetaDataCompat` but these are deprecated so should not get new uses.

If I put these deprecated members in SstFileMetaData, I can't use the `std::string& db_path = directory` trick without implementing full copy and move constructors/operators on all the members.

Reply:
If we think SstFileMetaData is a data structure to deprecate, we should worry less about paying this extra memory copy. I would vote for a little bit inefficiency in metadata path, against complicated inheritance.